codelabs/codelab-02/codelab2README.md

```diff
@@ -0,0 +1,501 @@
+
+```
+$ time aws s3api get-object —bucket cmsc389l-colink —key hello_world.txt out.txt
+...
+aws s3api get-object —bucket cmsc389l-colink —key hello_world.txt out.txt 0.29s user 0.05s system 41% cpu 0.817 total
+```
+
+Instead, if you were to move these files into Glacier, then you would need to trigger a transfer job which would move files back into an S3 bucket. This process will usually take hours.
+#### S3 CLI
+
+The tool that you've used so far, `aws s3api`, is one of two AWS CLI tools for interacting with S3. The other, `aws s3`, is a higher-level wrapper on top of `s3api` that provides typical directory traversal operations.
+
+For example, you can `ls` a directory on S3:
+
+```
+$ aws s3 ls s3://cmsc389l-colink
+PRE dir1/
+PRE dir2/
+2017-09-13 11 (tel:2017091311):44:17 13 hello_world.txt
+2017-09-13 12 (tel:2017091312):24:14 25165824 output.dat
+```
+
+It provides a number of operations:
+
+```
+cp ls mb mv presign rb rm sync website
+```
+
+I recommend that you play around with this tool a bit more to see how it compares with `s3api`. As you will see, it is much more limited than `s3api`, but it is quite useful for when you need to move files between your local computer and an S3 bucket.
+
+### Assignment
+
+Your assignment for this codelab is to re-create part of the `s3 sync` command using boto3 in Python. You'll be able to use this script to quickly copy files and directories into S3.
+
+**Note**: You must use boto3, not the `s3 sync` command. You will not get full credit if you just write a wrapper on top of the `sync` command.
+
+As an example of how this script works, say you had a directory like so:
+
+```
+a.txt
+folder1/
+b.txt
+c.txt
+folder2/
+folder3/
+d.txt
+```
+
+Then here are a few examples of the resulting state in S3:
+
+##### Example 1
+
+```
+$ python upload.py a.txt —bucket cmsc389l-colink
+```
+
+```
+a.txt
+```
+
+##### Example 2
+
+```
+$ python upload.py a.txt —bucket cmsc389l-colink —destination z.txt
+```
+
+```
+z.txt
+```
+
+##### Example 3
+
+```
+$ python upload.py . —bucket cmsc389l-colink
+```
+
+```
+a.txt
+folder1/b.txt
+folder1/c.txt
+folder2/folder3/d.txt
+```
+
+##### Example 4
+
+```
+$ python upload.py folder1 —bucket cmsc389l-colink —destination uploads
+```
+
+```
+uploads/b.txt
+uploads/c.txt
+```
+
+The `upload.py` file in the `code/` directory includes some boilerplate to get you started.
+
+Like the previous codelab, you only need to pass the public tests to get full credit.
+
+The tests can be run like so:
+
+```
+$ python test.py
+```
+
+Note that if you kill the testing program, it won't clean up after itself. If you end up with extra testing buckets, you can quickly remove them with the following command:
+
+```
+$ aws s3 ls | grep "test-bucket" | cut -f3 -d ' ' | xargs -I {} aws s3 rb 's3://{}'
+```
+
+Feel free to install any packages you would like to use. Just make sure to add them to the `requirements.txt` file (run `pip freeze > requirements.txt`).
```

Comment:
Swap this with pipenv install command, and replace other references to `requirements.txt`